### PR TITLE
Pushdown dereference

### PR DESCRIPTION
**Pull request by zhenxiao**
== RELEASE NOTES ==
```
General Changes
* push down dereference expression
```

### CODE REVIEW

**Comment by highker** [comment]
Quick question on this PR before starting review on this: what is the difference between `pushdown_dereference_enabled` vs `pushdown_subfields_enabled`?

**Comment by zhenxiao** [comment]
I think `pushdown_subfields_enabled ` is pushing down all subfields into TableScan. 
`pushdown_dereference_enabled ` is to pushdown dereferences on top of TableScan, so that connectors could optimize nested column pruning at their needs, e.g. with subfields pushdown, we will generate one block for `msg.[a, b, c]`, instead of separate blocks for `msg.a`, `msg.b`, and `msg.c`, which will hurt performance due to lack of lazy blocks, discussions in https://github.com/prestodb/presto/issues/14517. With `pushdown_dereference_enabled ` Hive connector and Elasticsearch connector could optimize dereferences pruning

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 1]
```diff
@@ -365,6 +366,13 @@ public PlanOptimizers(
                                 new TransformCorrelatedSingleRowSubqueryToProject())),
                 new CheckSubqueryNodesAreRewritten(),
                 predicatePushDown,
+                new IterativeOptimizer(
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:369]: Does this rule have to live above `TranslateExpressions`? If possible, we can have it between `TranslateExpressions` and `AddExchange`

Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:369]: yep, intended to be above `TranslateExpressions`. Use OriginalExpressionUtils.

Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:369]: But why? Any other specific rule it has to run after?

Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:369]: ah...this work started long time ago, before `TranslateExpressions` introduced. When porting it to current code,  the implementation is based on `OriginalExpressionUtils` `cast`

Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:373]: Can we have this patch written in `RowExpression` only? Check #14914

**Comment by zhenxiao** [comment]
hi @highker move dereference pushdown below `TranslateExpressions`, and it is working with `RowExpression` only
could you please take a review?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.DEREFERENCE;
+import static com.facebook.presto.sql.planner.ExpressionExtractor.extractExpressionsNonRecursive;
+import static com.facebook.presto.sql.planner.VariablesExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.AssignmentUtils.identityAssignments;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.facebook.presto.sql.planner.plan.Patterns.semiJoin;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.planner.plan.Patterns.unnest;
+import static com.facebook.presto.sql.tree.ExpressionTreeRewriter.rewriteWith;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:68]: unused

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+            super(FilterNode.class);
+        }
+
+        @Override
+        protected FilterNode rewrite(Context context, FilterNode node, BiMap<SpecialFormExpression, VariableReferenceExpression> expressions)
+        {
+            PlanNode source = node.getSource();
+
+            Map<VariableReferenceExpression, RowExpression> dereferencesMap = expressions.inverse().entrySet().stream()
+                    .collect(toImmutableMap(Map.Entry::getKey, entry -> entry.getValue()));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:177]: Map.Entry::getValue

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+            return project().with(source().matching(targetPattern.capturedAs(targetCapture)));
+        }
+
+        @Override
+        public Result apply(ProjectNode node, Captures captures, Context context)
+        {
+            N child = captures.get(targetCapture);
+            Map<SpecialFormExpression, VariableReferenceExpression> allDereferencesInProject = getDereferenceSymbolMap(node.getAssignments().getExpressions(), context, metadata);
+
+            Set<VariableReferenceExpression> childSourceSymbols = child.getSources().stream()
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:265]: childSourceVariables

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, N targetNode, BiMap<SpecialFormExpression, VariableReferenceExpression> expressions)
+        {
+            PlanNode source = getOnlyElement(targetNode.getSources());
+
+            Map<VariableReferenceExpression, RowExpression> dereferencesMap =
+                    expressions.inverse().entrySet().stream()
+                            .collect(toImmutableMap(Map.Entry::getKey, entry -> entry.getValue()));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:320]: same

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+        PushDownDereferenceThroughProject()
+        {
+            super(project());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, ProjectNode projectNode, BiMap<SpecialFormExpression, VariableReferenceExpression> expressions)
+        {
+            Map<VariableReferenceExpression, RowExpression> dereferencesMap = expressions.inverse().entrySet().stream()
+                    .collect(toImmutableMap(Map.Entry::getKey, entry -> entry.getValue()));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:436]: same

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+        PushDownDereferenceThroughUnnest()
+        {
+            super(unnest());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, UnnestNode unnestNode, BiMap<SpecialFormExpression, VariableReferenceExpression> expressions)
+        {
+            Map<VariableReferenceExpression, RowExpression> dereferencesMap = expressions.inverse().entrySet().stream()
+                    .collect(toImmutableMap(Map.Entry::getKey, entry -> entry.getValue()));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:460]: same

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 8]
```diff
@@ -401,7 +385,7 @@ protected Result pushDownDereferences(Context context, JoinNode joinNode, BiMap<
                             .addAll(leftNode.getOutputVariables())
                             .addAll(rightNode.getOutputVariables())
                             .build(),
-                    joinNode.getFilter(),
+                    joinNode.getFilter().map(expression -> replaceDereferences(expression, expressions)),
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:388]: why don't we have this before?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.unnest;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.SortItem.NullOrdering.LAST;
+import static com.facebook.presto.sql.tree.SortItem.Ordering.ASCENDING;
+
+public class TestDereferencePushDown
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java:45]: Let's have this to match the name `TestPushDownDereferences`. Or you wanna call the other `DeferencePushDown`. Personally I prefer the latter.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 8 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:388]: not needed, we already have ExtractFromJoin. let me fix

**Comment by zhenxiao** [comment]
thank you, @highker 
get comments addressed

**Pull request merged**